File: src/typical90/typical90_aq/main.jl
using DataStructures: deque

Pos = Tuple{Int, Int}

function bfs(g::Matrix{Bool}, w::Int, h::Int, from::Pos, to::Pos)::Int
    inf::Int = 10 ^ 6
    dx::Vector{Int} = [1, 0, -1, 0]
    dy::Vector{Int} = [0, 1, 0, -1]
    costs::Array{Int, 3} = fill(inf, (4, w, h))
    # pos, cost, dire
    q = Deque{Tuple{Pos, Int}}()

    for i in 1:4
        costs[i, from[1], from[2]] = 0
        push!(q, (from, i))
    end

    while !isempty(q)
        (x, y), pre_i = popfirst!(q)
        @inbounds for i in 1:4
            nx, ny = x + dx[i], y + dy[i]

            if !(1 <= nx <= w) || !(1 <= ny <= h)
                continue
            end

            if g[nx, ny]
                continue
            end

            cost = costs[pre_i, x, y] + ifelse(i == pre_i, 0, 1)
            if cost < costs[i, nx, ny]
                costs[i, nx, ny] = cost
                if i == pre_i # same dire
                    pushfirst!(q, ((nx, ny), i))
                else # differnt dire
                    push!(q, ((nx, ny), i))
                end
            end
        end
    end

    return minimum(costs[:, to[1], to[2]])
end

ints = () -> readline() |> split .|> e -> parse(Int, e)

function main()
    h, w = ints()
    rs, cs = ints()
    rt, ct = ints()
    g::Matrix{Bool} = zeros((w, h))
    for i in 1:h
        s = readline()
        for j in 1:w
            g[j, i] = s[j] != '.'
        end
    end

    ans = bfs(g, w, h, (cs, rs), (ct, rt))
    println(ans)
end

main()

File: src/typical90/typical90_aq/main.py
from collections import deque

h, w = list(map(int, input().split()))
rs, cs = list(map(int, input().split()))
rt, ct = list(map(int, input().split()))
g = []
for i in range(h):
    g.append(input())


def dire2pos(dire):
    if dire == 0:
        return [0, -1]
    elif dire == 1:
        return [1, 0]
    elif dire == 2:
        return [0, 1]
    else:
        return [-1, 0]

def bfs(g, w, h, start, goal):
    inf = 10 ** 6
    dires = list(range(4))
    costs = [[[inf, inf, inf, inf] for _ in range(w)] for _ in range(h)]

    # pos, dire
    q = deque()
    costs[start[1]][start[0]] = [0, 0, 0, 0]
    for d in dires:
        q.append((start, d))

    while q:
        pos, predire = q.popleft()

        for dire in dires:
            nd = dire2pos(dire)
            newpos = (pos[0] + nd[0], pos[1] + nd[1])

            if not (0 <= newpos[0] < w) or not (0 <= newpos[1] < h):
                continue
            # print(newpos)
            if g[newpos[1]][newpos[0]] == "#":
                continue

            cost = costs[pos[1]][pos[0]][predire] + (1 if dire != predire else 0)
            if cost < costs[newpos[1]][newpos[0]][dire]:
                costs[newpos[1]][newpos[0]][dire] = cost
                if dire == predire: # same dire
                    q.appendleft((newpos, dire))
                else: # differnt dire
                    q.append((newpos, dire))

    return min(costs[goal[1]][goal[0]])

print(bfs(g, w, h, (cs - 1, rs - 1), (ct - 1, rt - 1)))
